File: day-07/part-1/ayoub.py
from runners.python import Submission


class AyoubSubmission(Submission):

	def find_root(self, graph):
		for node in graph:
			has_parent = False
			for x in graph:
				if node in graph[x]:
					has_parent = True
					break
			if not has_parent:
				return node

	def run(self, s):
		graph = {}
		for line in s.split('\n'):
			parts = line.split()
			if len(parts) == 2:
				graph[parts[0]] = set()
			else:
				graph[parts[0]] = set(''.join(parts[3:]).split(','))
		return self.find_root(graph)

File: day-07/part-1/ayoub.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func findRoot(graph map[string][]string) string {
	parent := make(map[string]string)

	for node, children := range graph {
		for _, child := range children {
			parent[child] = node
		}
	}

	for node := range graph {
		if _, ok := parent[node]; !ok {
			return node
		}
	}
	return ""
}

func run(s string) string {
	graph := make(map[string][]string)
	for _, line := range strings.Split(s, "\n") {
		parts := strings.Split(line, " ")
		if len(parts) == 2 {
			graph[parts[0]] = []string{}
		} else {
			graph[parts[0]] = append(graph[parts[0]], strings.Split(strings.Join(parts[3:], ""), ",")...)
		}
	}
	return findRoot(graph)
}

func main() {
	fmt.Println(run(os.Args[1]))
}
